### PR TITLE
Add a PR template and issue template

### PR DESCRIPTION
**Pull request by OussamaSaoudi-db**
### What changes were proposed in this pull request?

This PR adds a pull request template and issues template to the repository.
The issues template is adapted from Apache Datafusion's issue template: https://github.com/apache/datafusion/.
This pull request template is adapted from Apache Spark's pull request template: https://github.com/apache/spark

### Why are the changes needed?

A pull request template encourages open source contributions to clearly state their purpose and approach. It explicitly calls out user facing changes and explains why they are needed. It also encourages all contributions to be well tested. In the event testing was difficult or impossible, this exposes testing pain points can be addressed.

Issues template encourages community members  and users to give all the details that would help bugs get fixed and feature requests implemented. Requiring more details from issues also helps newcomers to the project get more context to get started contributing.

### Does this PR introduce _any_ user-facing change?

No

### How was this change tested?

No way to test pull request templates or issue templates.

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  1. If you propose a new API or feature, clarify the use case for a new API or feature.
+  2. If you fix a bug, you can clarify why it is a bug.
+-->
+
+
+### Does this PR introduce _any_ user-facing change?
+<!--
+Note that it means *any* user-facing change including all aspects such as the documentation fix.
+If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
+If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
```
Comment (scottsand-db) [inline .github/pull_request_template.md:30]: Spark? :p

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
+<!--
+Thanks for sending a pull request!  Here are some tips for you:
+  1. If this is your first time, please read our contributor guidelines: https://github.com/delta-incubator/delta-kernel-rs/blob/main/CONTRIBUTING.md
+  2. Ensure you have added or run the appropriate tests for your PR
```
Comment (nicklan) [inline .github/pull_request_template.md:4]:
```suggestion
  2. Ensure you have added or run the appropriate tests for your PR and run `cargo fmt`.
```

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  2. If you fix a bug, you can clarify why it is a bug.
+-->
+
+
+### Does this PR introduce _any_ user-facing change?
+<!--
+Note that it means *any* user-facing change including all aspects such as the documentation fix.
+If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
+If no, write 'No'.
+-->
```
Comment (nicklan) [inline .github/pull_request_template.md:31]: This feels a bit "heavy".  Maybe something more like:
```suggestion
### Does this PR effect any _public_ APIs?
<!--
If yes, please ensure the `breaking-changes` label gets added by CI, and describe why the changes are needed.

Note that _new_ public APIs are not considered breaking.
-->
```

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+### How was this change tested?
+<!--
+If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
+If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
+If tests were not added, please describe why they were not added and/or why it was difficult to add.
+-->
+
+
+<!--
+Adapted from Apache Spark's pull request template: https://github.com/apache/spark
```
Comment (nicklan) [inline .github/pull_request_template.md:43]: Do we need this?

File: .github/ISSUE_TEMPLATE/feature_request.yml [hunk 5]
```diff
@@ -0,0 +1,26 @@
+name: Feature request
+description: Suggest an idea for this project
+labels: enhancement
+body:
+  - type: textarea
+    attributes:
+      label: Please describe what you are trying to do.
```
Comment (nicklan) [inline .github/ISSUE_TEMPLATE/feature_request.yml:7]:
```suggestion
      label: Please describe why this is necessary.
```

File: .github/ISSUE_TEMPLATE/feature_request.yml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+      placeholder: >
+        A clear and concise description of what the problem is, or what motivates the feature you are requesting
+  - type: textarea
+    attributes:
+      label: Describe the solution you'd like
```
Comment (nicklan) [inline .github/ISSUE_TEMPLATE/feature_request.yml:12]:
```suggestion
      label: Describe the functionality you are proposing.
```

File: .github/ISSUE_TEMPLATE/feature_request.yml [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+      placeholder: >
+        A clear and concise description of what you want to happen.
+  - type: textarea
+    attributes:
+      label: Describe alternatives you've considered
+      placeholder: >
+        A clear and concise description of any alternative solutions or features you've considered.
```
Comment (nicklan) [inline .github/ISSUE_TEMPLATE/feature_request.yml:19]:
```suggestion
        List alternative solutions or features you've considered.
```

File: .github/ISSUE_TEMPLATE/bug_report.yml [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  - type: textarea
+    attributes:
+      label: To Reproduce
+      placeholder: >
+        Steps to reproduce the behavior:
+  - type: textarea
+    attributes:
+      label: Expected behavior
+      placeholder: >
+        A clear and concise description of what you expected to happen.
```
Comment (nicklan) [inline .github/ISSUE_TEMPLATE/bug_report.yml:21]:
```suggestion
        What you expected to happen.
```

**Review by nicklan — CHANGES_REQUESTED** [review]
Thanks. Just a few wording suggestions.

File: .github/pull_request_template.md (hunk 4 above)
Comment (zachschuermann) [reply .github/pull_request_template.md:43]: i think this was originally my suggestion :) but i meant for us to just credit them in the PR description or something - don't need to have it in the actual template

File: .github/ISSUE_TEMPLATE/bug_report.yml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  - type: textarea
+    attributes:
+      label: Additional context
+      placeholder: >
+        Add any other context about the problem here.
+
+# Adapted from Apache Datafusion's issue template: https://github.com/apache/datafusion/
```
Comment (zachschuermann) [inline .github/ISSUE_TEMPLATE/bug_report.yml:28]: just credit in PR description :)
```suggestion
```

File: .github/ISSUE_TEMPLATE/feature_request.yml [hunk 10]
```diff
@@ -0,0 +1,26 @@
+name: Feature request
+description: Suggest an idea for this project
+labels: enhancement
+body:
+  - type: textarea
+    attributes:
+      label: Please describe what you are trying to do.
+      placeholder: >
+        A clear and concise description of what the problem is, or what motivates the feature you are requesting
```
Comment (zachschuermann) [inline .github/ISSUE_TEMPLATE/feature_request.yml:9]:
```suggestion
        A description of what the problem is, or what motivates the feature you are requesting
```

File: .github/ISSUE_TEMPLATE/feature_request.yml [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+  - type: textarea
+    attributes:
+      label: Describe the solution you'd like
+      placeholder: >
+        A clear and concise description of what you want to happen.
+  - type: textarea
+    attributes:
+      label: Describe alternatives you've considered
```
Comment (zachschuermann) [inline .github/ISSUE_TEMPLATE/feature_request.yml:17]: overkill?

File: .github/ISSUE_TEMPLATE/feature_request.yml [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+      placeholder: >
+        A clear and concise description of any alternative solutions or features you've considered.
+  - type: textarea
+    attributes:
+      label: Additional context
+      placeholder: >
+        Add any other context or screenshots about the feature request here.
+
+# Adapted from Apache Datafusion's issue template: https://github.com/apache/datafusion/
```
Comment (zachschuermann) [inline .github/ISSUE_TEMPLATE/feature_request.yml:26]:
```suggestion
```

File: .github/pull_request_template.md (hunk 2 above)
Comment (zachschuermann) [reply .github/pull_request_template.md:4]: probably also include high-level testing command? just say to run `cargo t --workspace --all-features` or something?

File: .github/pull_request_template.md [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+-->
+
+### What changes were proposed in this pull request?
+<!--
+Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue.
+If possible, please consider writing useful notes for better and faster reviews in your PR.
+-->
+
+
+### Why are the changes needed?
```
Comment (zachschuermann) [inline .github/pull_request_template.md:18]: can we remove this and just add a note in the comment above for 'what changes are proposed' to say 'Include why the changes are needed'? again just trying to streamline/simplify

File: .github/pull_request_template.md [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+Thanks for sending a pull request!  Here are some tips for you:
+  1. If this is your first time, please read our contributor guidelines: https://github.com/delta-incubator/delta-kernel-rs/blob/main/CONTRIBUTING.md
+  2. Ensure you have added or run the appropriate tests for your PR
+  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP] Your PR title ...'.
+  4. Be sure to keep the PR description updated to reflect all changes.
+  5. Please write your PR title to summarize what this PR proposes.
+  6. If possible, provide a concise example to reproduce the issue for a faster review.
+-->
+
+### What changes were proposed in this pull request?
```
Comment (zachschuermann) [inline .github/pull_request_template.md:11]:
```suggestion
### What changes are proposed in this pull request?
```

File: .github/pull_request_template.md [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+<!--
+Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue.
+If possible, please consider writing useful notes for better and faster reviews in your PR.
```
Comment (zachschuermann) [inline .github/pull_request_template.md:14]:
```suggestion
```

File: .github/pull_request_template.md [hunk 16]
```diff
@@ -0,0 +1,44 @@
+<!--
+Thanks for sending a pull request!  Here are some tips for you:
+  1. If this is your first time, please read our contributor guidelines: https://github.com/delta-incubator/delta-kernel-rs/blob/main/CONTRIBUTING.md
+  2. Ensure you have added or run the appropriate tests for your PR
+  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP] Your PR title ...'.
+  4. Be sure to keep the PR description updated to reflect all changes.
+  5. Please write your PR title to summarize what this PR proposes.
+  6. If possible, provide a concise example to reproduce the issue for a faster review.
```
Comment (zachschuermann) [inline .github/pull_request_template.md:8]: move `6.` to below?

File: .github/pull_request_template.md [hunk 17]
```diff
@@ -0,0 +1,44 @@
+<!--
+Thanks for sending a pull request!  Here are some tips for you:
+  1. If this is your first time, please read our contributor guidelines: https://github.com/delta-incubator/delta-kernel-rs/blob/main/CONTRIBUTING.md
+  2. Ensure you have added or run the appropriate tests for your PR
+  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP] Your PR title ...'.
+  4. Be sure to keep the PR description updated to reflect all changes.
+  5. Please write your PR title to summarize what this PR proposes.
```
Comment (zachschuermann) [inline .github/pull_request_template.md:7]:
```suggestion
```

File: .github/pull_request_template.md [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+Note that it means *any* user-facing change including all aspects such as the documentation fix.
+If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
+If no, write 'No'.
+-->
+
+
+### How was this change tested?
+<!--
+If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
+If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
```
Comment (zachschuermann) [inline .github/pull_request_template.md:37]:
```suggestion
If it was tested in a way different from regular unit tests, please clarify how you tested, ideally via a reproducible test documented in the PR description.
```

File: .github/pull_request_template.md [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+If tests were not added, please describe why they were not added and/or why it was difficult to add.
```
Comment (zachschuermann) [inline .github/pull_request_template.md:38]:
```suggestion
```

**Review by zachschuermann — CHANGES_REQUESTED** [review]
just lots of wording. but generally looks good - thanks! I'm mostly advocating for making it simpler (I think there is a balance between not enough/too much structure for these things)

**Comment by zachschuermann** [comment]
Also can we make the 'user-facing changes' section of the PR description hidden by default with a comment that says 'uncomment if user-facing changes'? that way we don't just have tons of PRs with 'user-facing changes?' 'No.'

File: .github/pull_request_template.md [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+The purpose of this section is to outline the changes, why they are needed, and how this PR fixes the issue.
+If the reason for the change is already explained clearly in an issue, then it does not need to be restated here.
+  1. If you propose a new API or feature, clarify the use case for a new API or feature.
+  2. If you fix a bug, you can clarify why it is a bug.
+  3. If possible, provide a concise example to reproduce the issue for a faster review.
+-->
+
+<!--
+Uncomment this section if there are any changes affecting public APIs:
+### Does this PR affect any _public_ APIs?
```
Comment (zachschuermann) [inline .github/pull_request_template.md:21]:
```suggestion
### This PR affects the following public APIs
```

File: .github/pull_request_template.md [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+
+If yes, please ensure the `breaking-changes` label gets added by CI, and describe why the changes are needed.
```
Comment (zachschuermann) [inline .github/pull_request_template.md:23]: Isn't required right? New public API could mean no breaking changes? Maybe we just say 'if there are breaking changes please ensure the label blah blah. note only _new_ public APIs are not considered breaking changes.

File: .github/pull_request_template.md [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+
+Note that _new_ public APIs are not considered breaking.
+-->
+
+
+### How was this change tested?
+<!--
+If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
```
Comment (zachschuermann) [inline .github/pull_request_template.md:31]:
```suggestion
Please make sure to add test cases that check the changes thoroughly including negative and positive cases if possible.
```

File: .github/pull_request_template.md [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+-->
+
+### What changes are proposed in this pull request?
+<!--
+Please clarify what changes you are proposing and why the changes are needed.
+The purpose of this section is to outline the changes, why they are needed, and how this PR fixes the issue.
+If the reason for the change is already explained clearly in an issue, then it does not need to be restated here.
+  1. If you propose a new API or feature, clarify the use case for a new API or feature.
+  2. If you fix a bug, you can clarify why it is a bug.
+  3. If possible, provide a concise example to reproduce the issue for a faster review.
```
Comment (zachschuermann) [inline .github/pull_request_template.md:16]: should this be here or in the issue template?

File: .github/pull_request_template.md [hunk 24]
```diff
@@ -0,0 +1,33 @@
+<!--
+Thanks for sending a pull request!  Here are some tips for you:
+  1. If this is your first time, please read our contributor guidelines: https://github.com/delta-incubator/delta-kernel-rs/blob/main/CONTRIBUTING.md
+  2. Ensure you have added or run the appropriate tests for your PR and run `cargo fmt`.
```
Comment (zachschuermann) [inline .github/pull_request_template.md:4]: can we add a quick one-liner for testing? Maybe something to the effect of "run `cargo t --all-features --all-targets` to get started testing"

**Review by zachschuermann — COMMENTED** [review]
nice - few comments :)

File: .github/pull_request_template.md (hunk 23 above)
Comment (OussamaSaoudi-db) [reply .github/pull_request_template.md:16]: Yeah this was part of spark one, but you're right that it makes more sense in an issue

File: .github/pull_request_template.md [hunk 25]
```diff
@@ -0,0 +1,33 @@
+<!--
+Thanks for sending a pull request!  Here are some tips for you:
+  1. If this is your first time, please read our contributor guidelines: https://github.com/delta-incubator/delta-kernel-rs/blob/main/CONTRIBUTING.md
+  2. Run `cargo t --all-features --all-targets` to get started testing, and run `cargo fmt`.
+  3. Ensure you have added or run the appropriate tests for your PR.
+  4. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP] Your PR title ...'.
+  5. Be sure to keep the PR description updated to reflect all changes.
+-->
+
+### What changes are proposed in this pull request?
```
Comment (zachschuermann) [inline .github/pull_request_template.md:10]: nit: make it H2?

**Review by zachschuermann — APPROVED** [review]
lgtm i vote we merge and iterate

**Review by nicklan — APPROVED** [review]
lgtm! thanks
